packages/runtime/src/decorators/middlewares.ts

```diff
@@ -0,0 +1,125 @@
+import 'reflect-metadata';
+export type ExpressMiddleware = (req: any, res: any, next: any) => Promise<any>;
+export type KoaMiddleware = (ctx: any, next: any) => Promise<any>;
+export type HapiMiddlewareBase = (request: any, h: any) => Promise<any>;
+export type HapiMiddlewareSimple = HapiMiddlewareBase | { method: HapiMiddlewareBase; assign?: string; failAction?: HapiMiddlewareBase | string };
+export type HapiMiddleware = HapiMiddlewareSimple | HapiMiddlewareSimple[];
+
+export type Middlewares = {
+  express?: ExpressMiddleware[];
+  koa?: KoaMiddleware[];
+  hapi?: HapiMiddleware[];
+};
+
+const TSOA_EXPRESS_MIDDLEWARES = Symbol('tsoa:expressMiddlewares');
+const TSOA_KOA_MIDDLEWARES = Symbol('tsoa:koaMiddlewares');
+const TSOA_HAPI_MIDDLEWARES = Symbol('tsoa:hapiMiddlewares');
+
+/**
+ * Helper function to create a decorator
+ * that can act as a class and method decorator.
+ * @param fn a callback function that accepts
+ *           the subject of the decorator
+ *           either the constructor or the
+ *           method
+ * @returns
+ */
+function decorator(fn: (value: any) => void) {
```

Comment:
TS has a MethodDecorator and ParameterDecorator type.
While we have historically not been very diligent, I think it makes sense to type the returns if it's not too complicated

Reply:
👍  actually I just typed `decorator` to return `ClassDecorator & MethodDecorator` and it works (not sure how `typescript` is able to understand that, though)